src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs

```diff
@@ -262,51 +281,52 @@ public static void AddProviderTransparent(TypeDescriptionProvider provider, obje
-                // this could infinitely recurse.
-                s_defaultProviders[type] = null;
-            }
+                        var providerAttr = type.GetCustomAttributes<TypeDescriptionProviderAttribute>(false)
+                            .SingleOrDefault();
+                        bool providerAdded = false;
 
-            // Always use core reflection when checking for
-            // the default provider attribute. If there is a
-            // provider, we probably don't want to build up our
-            // own cache state against the type. There shouldn't be
-            // more than one of these, but walk anyway. Walk in
-            // reverse order so that the most derived takes precidence.
-            object[] attrs = type.GetCustomAttributes(typeof(TypeDescriptionProviderAttribute), false);
-            bool providerAdded = false;
-            for (int idx = attrs.Length - 1; idx >= 0; idx--)
-            {
-                TypeDescriptionProviderAttribute pa = (TypeDescriptionProviderAttribute)attrs[idx];
-                Type? providerType = Type.GetType(pa.TypeName);
-                if (providerType != null && typeof(TypeDescriptionProvider).IsAssignableFrom(providerType))
-                {
-                    TypeDescriptionProvider prov = (TypeDescriptionProvider)Activator.CreateInstance(providerType)!;
-                    AddProvider(prov, type);
-                    providerAdded = true;
+                        if (providerAttr != null)
+                        {
+                            Type? providerType = Type.GetType(providerAttr.TypeName);
+                            if (providerType != null && typeof(TypeDescriptionProvider).IsAssignableFrom(providerType))
+                            {
+                                TypeDescriptionProvider provider = (TypeDescriptionProvider)Activator.CreateInstance(providerType)!;
+                                AddProvider(provider, type);
+                                providerAdded = true;
+                            }
+                        }
+
+                        // If we did not add a provider, check the base class.
+                        if (!providerAdded)
+                        {
+                            Type? baseType = type.BaseType;
+                            if (baseType != null && baseType != type)
+                            {
+                                CheckDefaultProvider(baseType);
+                            }
+                        }
+
+                        resetEventForAdd.Set();
+                    }
                 }
             }
-
-            // If we did not add a provider, check the base class.
-            if (!providerAdded)
+            else
             {
-                Type? baseType = type.BaseType;
-                if (baseType != null && baseType != type)
+                if (context.OwnerId != Environment.CurrentManagedThreadId)
```

Comment:
So we get here if `_processedTypes` has the type but we need to wait for the above to finish on the other thread that is adding the value?

Is this necessary? The AddProvider() method is already thread-safe for adding to `s_providerTable`

Reply:
Yes, this code is necessary. Without it `GetConverterWithAddProvider_ByMultithread_Success` is failed